### ISSUE TITLE
qwen4_exp: Qwen3.8-Flash-Next MLX 4-bit checkpoint loads cleanly on main but generates garbage

### ISSUE THREAD

**Issue by yobuce** [Blaizzy/mlx-vlm#2041]
`Vontra/Qwen3.8-Flash-Next-MLX-4bit` loads cleanly on `main` — every weight maps, no warnings — but generation is deterministic garbage. The same checkpoint files produce coherent text on the superseded #2028 branch, so I don't think it's the checkpoint.

Raising it because the failure is silent: nothing in the load path complains, and `mlx_vlm/tests/test_qwen4_exp.py` passes 8/8.

## Repro

```sh
pip install "git+https://github.com/Blaizzy/mlx-vlm@main"   # c2a67b63
hf download Vontra/Qwen3.8-Flash-Next-MLX-4bit --local-dir Q38FN-4bit

python -m mlx_vlm.generate --model Q38FN-4bit \
  --prompt "Write a short paragraph explaining what Gated DeltaNet is and why it helps long-context inference." \
  --max-tokens 128 --temperature 0.0 --verbose
```

Output:

```
  
4<|im_start|>
###

Prompt: 31 tokens, 1.070 tokens-per-sec
Generation: 7 tokens, 33.636 tokens-per-sec
Peak memory: 111.835 GB
```

Seven tokens and it stops. Byte-identical across runs and across `505267ca` / `c2a67b63`, so it's deterministic rather than a sampling artifact.

For contrast, the same directory on `JJJYmmm:add_qwen4exp` (#2028) — plus two small compat patches, see below — gives:

> Gated DeltaNet is an advanced recurrent neural network architecture that integrates the delta rule (a fast, parallelizable update mechanism) with a gating mechanism to manage information flow within the hidden state. It helps long-context inference by enabling efficient, linear-time processing of sequences, avoiding the quadratic computational complexity and memory overhead associated with traditional Transformers' self-attention mechanisms. […]

at 33.9 tok/s decode, 658 tok/s prefill on a 2,585-token prompt.

## What I ruled out

- **Checkpoint integrity** — 22/22 shards, 3,671 tensors, on-disk bytes exactly equal `metadata.total_size` (111,578,339,151).
- **Mixed precision hiding in the checkpoint** — all 998 quantized modules have `weight_cols / scales_cols == 4.0`, i.e. uniformly 4-bit at group 32, matching `config.json`'s `{"group_size": 32, "bits": 4, "mode": "affine"}`. No per-module exceptions.
- **Weight mapping** — `load_weights` is strict here and raised nothing, so the key sets match exactly after `sanitize`.
- **n-gram hash constants** — `_build_layer_multipliers` and the `_find_nth_prime_after` head tables on `main` produce values identical to the #2028 branch: multipliers `[23703573157769, 20109073645365, 8052911324071]`, head sizes `[20000003, 20000023, 20000033, 20000047, …]`, offsets `[0, 20000003, 40000026, 60000059, …]`, total 320,001,446 → padded 320,001,536 = 128 × 2,500,012, which is exactly the shard layout in the checkpoint.
- **`conv1d` axis handling** — all 37 `conv1d.weight` tensors are `(10240, 4, 1)`, so `shape[-1] != 1` is false and the `moveaxis` in `sanitize` correctly no-ops.
- **Stale checkout** — reproduced on `c2a67b63` with `test_qwen4_exp.py` from the same commit passing 8/8. (An earlier 1/8 failure on my side was my own version skew, not a real failure.)

So the weights are all present, correctly shaped, correctly dequantized as far as I can tell, and the derived constants agree — which points at the forward pass rather than loading.

## Question

Is this checkpoint layout expected to work on `main`? It was converted before #2032 landed, with the #2028 branch's converter. If community conversions need redoing against `main`'s converter, that's worth a line in `models/qwen4_exp/README.md` — several 4-bit/8-bit MLX conversions of this model were published on the Hub in the hours around the merge, and this failure mode gives users no hint that the checkpoint is the problem.

Happy to run diagnostics on my side — I have the model resident and can dump intermediate activations, per-layer norms, or first-token logits against the #2028 implementation for the same input if that would help localize it.

## Environment

- M4 Max, 128 GB, macOS 26.5.2, `iogpu.wired_limit_mb=122880`
- `mlx 0.32.2`, `mlx-vlm` @ `c2a67b63`, `transformers 5.16.1`, Python 3.11.14
- Peak memory 111.8 GB on `main` vs 135.9 GB on #2028 — the sharded n-gram embedding is a real improvement, worth keeping

<sub>Two patches were needed for #2028 to read this checkpoint at all, in case they're useful context: its `config.py` rejects `layer_types: "full_attention"` (`main` already maps it to `qwen_sparse_attention`), and its `merge_ngram_embedding_shards` only matches `.shard_N.weight`, so quantized shards leave `.scales`/`.biases` unmatched (`main`'s `ShardedEmbedding` handles this correctly).</sub>

**Comment by gtonic** [comment]
Diagnosis: `qwen4_exp` reads the RMSNorm weights with the **opposite centering convention** to the one the checkpoints store. This is sufficient on its own to produce the deterministic garbage in this report.

## The divergence

`mlx_vlm/models/qwen4_exp/language.py:576-597`:

```python
class Qwen4ExpRMSNorm(nn.Module):
    """Qwen4 RMSNorm, whose checkpoint weights are centered at zero."""
    ...
        self.weight = mx.zeros(dim)                  # :585
    ...
        y = y * (1.0 + weight.astype(mx.float32))    # :596
```

`qwen4_exp/qwen4_exp.py::sanitize` does not touch any norm weight, so whatever is on disk goes straight into that `1.0 + w`.

But the checkpoints store **standard gamma centered at 1.0**. Measured on `Vontra/Qwen3.8-Flash-Next-MLX-oQ2`, 51 norm tensors across the 8 suffixes that `Qwen4ExpRMSNorm` serves:

```
    mean      [min,      max]     tensor
   1.360  [ 0.254,   1.727]  layers.11.self_attn.k_norm.weight
   1.283  [ 0.656,   1.594]  layers.3.self_attn.q_norm.weight
   0.963  [ 0.602,   1.023]  layers.3.self_attn.indexer.q_layernorm.weight
   0.936  [-4.938,   7.625]  layers.0.attn_hyper_connection.hc_norm.weight
   0.890  [-1.188,   4.625]  layers.0.mlp_hyper_connection.hc_norm.weight
   3.750  [ 0.633,  13.375]  hyper_connection_mixer.hc_norm.weight

  mean of means over all 51 = 1.128
```

Every gain is therefore inflated by roughly `1 + w` instead of `w` — q_norm 1.28 becomes 2.28, so attention logits scale ~4×. Worse, `attn_hyper_connection.hc_norm` spans −4.94…+7.63, so **every channel with `w ∈ (−1, 0)` flips sign**. That happens twice per layer across all 48 layers, in the module that mixes the hyper-connection residual streams.

## Internal control, same file

`Qwen4ExpRMSNormGated` (`language.py:600-615`) initialises `mx.ones(dim)` and calls `mx.fast.rms_norm(x, self.weight, eps)` — plain standard gamma. It reads `linear_attn.norm.weight`, which measures **mean 0.967**, i.e. the same distribution as `indexer.q_layernorm.weight` at 0.963 that the zero-centered class reads. One file, one checkpoint, two mutually exclusive conventions. Exactly one can be right.

`qwen4_exp` is also the only model in the family that opted out: `qwen3_5/qwen3_5.py:150,162-166`, `qwen3_5_moe/qwen3_5_moe.py:31,85-88` and `minicpmv4_6/minicpmv4_6.py:431,486-487` all do the `+1` shift at load time behind `should_shift_norm_weights`, and keep the module itself a plain RMSNorm.

#2028 does the same thing and is consistent with the checkpoints — `Qwen4ExpGroupRMSNorm` uses `mx.ones` and `normed * self.weight`, with the shift in `sanitize()` behind the already-converted guard, over exactly the eight suffixes `Qwen4ExpRMSNorm` is instantiated for:

```python
NORM_WEIGHT_SUFFIXES = (".q_norm.weight", ".k_norm.weight", ".q_layernorm.weight",
                        ".k_layernorm.weight", ".hc_norm.weight", ".norm_key.weight",
                        ".norm_query.weight", ".norm_conv.weight")
```

## A/B

Monkeypatching the offset away — nothing else changed, `--temperature 0.0`, identical prompt and command:

| | output |
|---|---|
| main as-is | `C,.com.com (em of\n\nP" $\n (,... (` |
| `y * w` instead of `y * (1 + w)` | `The main subject of this request is to be a single, no real` |

Token salad becomes well-formed English.

## Caveat, stated plainly

The residual output is still wrong — fluent but semantically empty (`"The capital of France is"` → `"...a common term for a human's ... a human's"`). That is **not** evidence against the above; the checkpoint I measured on carries a quality hold from its own producer ("live testing found incoherent output, do not use it as a quality checkpoint") and leaves `lm_head` over a 248,320-row vocabulary at 2 bits. So I can show the norm convention is inverted and that fixing it changes the failure class, but I cannot show a coherent model on this checkpoint.

Since this report is on a 4-bit checkpoint, worth confirming the same on yours — it needs no model load and no GPU:

```python
import glob, mlx.core as mx
sfx = (".hc_norm.weight", ".q_norm.weight", ".k_norm.weight", ".norm_key.weight",
       ".norm_query.weight", ".norm_conv.weight", ".q_layernorm.weight", ".k_layernorm.weight")
n = 0
for f in sorted(glob.glob("<model_dir>/*.safetensors")):
    for k, v in mx.load(f).items():
        if v.ndim == 1 and k.endswith(sfx):
            v = v.astype(mx.float32)
            print(f"{k:70s} mean={float(v.mean()):+.4f}")
            n += 1
            if n >= 12: raise SystemExit
```

mean ≈ +1.0 confirms it; mean ≈ 0.0 would mean the convention is right for that file and the cause is elsewhere.

## Suggested shape of a fix

Give `qwen4_exp/qwen4_exp.py::sanitize` the same `should_shift_norm_weights` / `should_offset_norm_weight` treatment the rest of the family uses, and revert `Qwen4ExpRMSNorm` to a plain `mx.fast.rms_norm` over `mx.ones`. That makes `qwen4_exp` agree with `qwen3_5`, `qwen3_5_moe` and `minicpmv4_6` on the on-disk convention, and loads both raw-HF and MLX-converted checkpoints correctly.

Unrelated, noted in passing since it may explain the 1.07 tok/s prefill in this report: `ShardedEmbedding.__call__` (`language.py:963-964`) does `mx.eval(flat)` followed by `.tolist()` — a host sync per PLE layer per forward step, then a Python loop that rescans all indices once per touched shard.

**Comment by Lazarus-931** [comment]
hey @gtonic @yobuce, model conversion was wrong, which is why i think pushing code for this specific model is wrong, so I converted it to proper format:   https://huggingface.co/mlx-community/Qwen3.8-Flash-Next-4bit/settings

**Comment by gtonic** [comment]
@Lazarus-931 is right, and my suggested fix above was wrong. Correcting it, since I don't want a bad recommendation sitting on an open issue.

I probed the norm tensors in [`mlx-community/Qwen3.8-Flash-Next-4bit`](https://huggingface.co/mlx-community/Qwen3.8-Flash-Next-4bit) by range-reading them out of the shards, and compared against the same tensors in the checkpoint I had measured:

| tensor | new conversion | the checkpoint I measured | diff |
|---|---|---|---|
| `layers.0.attn_hyper_connection.hc_norm.weight` | −0.064 | +0.936 | **−1.000** |
| `layers.3.self_attn.q_norm.weight` | +0.283 | +1.283 | **−1.000** |
| `layers.3.self_attn.indexer.q_layernorm.weight` | −0.037 | +0.963 | **−1.000** |
| `layers.11.self_attn.k_norm.weight` | +0.360 | +1.360 | **−1.000** |

Exactly −1.000 on every one. Same weights, zero-centered — which is what `Qwen4ExpRMSNorm`'s `1.0 + w` expects.

So the observation in my earlier comment holds (the checkpoint and the code disagreed on centering, and that disagreement is sufficient to produce the garbage), but the attribution was backwards: the checkpoints in circulation had the `+1` baked in at conversion time, and `main` is correct as written. **Please disregard the "suggested shape of a fix" section** — changing `Qwen4ExpRMSNorm` would break correctly-converted checkpoints.

Worth keeping from it, if anything: the 5-second disk probe is a cheap way to tell the two checkpoint families apart, and adopting the family's `should_shift_norm_weights` guard would let `qwen4_exp` load both rather than only one. That's a robustness question, not a bug — and given a correct conversion now exists, probably not worth the complexity.

The perf note stands independently: `ShardedEmbedding.__call__` (`language.py:963-964`) does `mx.eval(flat)` then `.tolist()`, a host sync per PLE layer per forward step.

Apologies for the noise, and thanks for doing the conversion.

**Comment by Lazarus-931** [comment]
closing as addressed by: https://github.com/Blaizzy/mlx-vlm/issues/2041#issuecomment-5486592276
